app/helpers/automated_tests_helper.rb

```diff
@@ -480,6 +486,289 @@ def parse_test_output(repo_assignment_dir, assignment, logfile, data)
+  def self.has_permission?()
+    if @current_user.admin?
+      true
+    elsif @current_user.ta?
+      true
+    elsif @current_user.student?
+      # Make sure student belongs to this group
+      if not @current_user.accepted_groupings.include?(@grouping)
+        # TODO: show the error to user instead of raising a runtime error
+        raise I18n.t("automated_tests.not_belong_to_group")
+      end
+      #can skip checking tokens if we have unlimited
+      if @grouping.assignment.unlimited_tokens
+        return true
+      end
+      t = @grouping.token
+      if t == nil
+        raise I18n.t('automated_tests.missing_tokens')
+      end
+      if t.tokens > 0
+        t.decrease_tokens
+        true
+      else
+        # TODO: show the error to user instead of raising a runtime error
+        raise I18n.t("automated_tests.missing_tokens")
+      end
+    end
+  end
+
+
+  # Perform a job for automated testing. This code is run by
+  # the Resque workers - it should not be called from other functions.
+  def self.perform(grouping_id, call_on)
+    # Pick a server, launch the Test Runner and wait for the result
+    # Then store the result into the database
+
+    #@submission = Submission.find(submission_id)
+    @grouping = Grouping.find(grouping_id)
+    @assignment = @grouping.assignment
+    @group = @grouping.group
+    @repo_dir = File.join(MarkusConfigurator.markus_config_automated_tests_repository, @group.repo_name)
+
+    # @list_of_servers = MarkusConfigurator.markus_ate_test_server_hosts.split(' ')
+
+    # while true
+    #   @test_server_id = choose_test_server()
+    #   if @test_server_id >= 0
+    #     break
+    #   else
+    #     sleep 5               # if no server is available, sleep for 5 second before it checks again
+    #   end
+    # end
+
+    stderr, result, status = launch_test(@assignment, @repo_dir, call_on)
+
+    if !status
+      #for debugging any errors in launch_test
+      # server_id = @test_server_id
+      assignment = @assignment
+      repo_dir = @repo_dir
+      m_logger = MarkusLogger.instance
+
+
+      src_dir = File.join(repo_dir, assignment.repository_folder)
+
+      # Get test_dir
+      test_dir = File.join(MarkusConfigurator.markus_config_automated_tests_repository, assignment.repository_folder)
+
+      # Get the name of the test server
+      # server = @list_of_servers[server_id]
+      server = 'localhost'
+
+      # Get the directory and name of the test runner script
+      test_runner = MarkusConfigurator.markus_ate_test_runner_script_name
+
+      # Get the test run directory of the files
+      run_dir = MarkusConfigurator.markus_ate_test_run_directory
+
+
+      m_logger.log("error with launching test, error: #{stderr} and status: #{status}\n src_dir: #{src_dir}\ntest_dir: #{test_dir}\nserver: #{server}\ntest_runner: #{test_runner}\nrun_dir: #{run_dir}",MarkusLogger::ERROR)
+
+      # TODO: handle this error better
+      raise "error"
```

Comment:
Prefer single-quoted strings when you don't need string interpolation or special symbols.